File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.cc
#include <iostream>
#include <cmath>
#include <string>
using namespace std;

// 재귀적으로 가운데 공백을 만드는 함수
void makeCantor(string &s, int start, int end) {
    if (end - start < 3) return;

    int third = (end - start) / 3;

    // 가운데 구간을 공백으로 설정
    for (int i = start + third; i < start + 2 * third; i++) {
        s[i] = ' ';
    }

    // 좌, 우 구간에 대해 재귀 호출
    makeCantor(s, start, start + third);
    makeCantor(s, start + 2 * third, end);
}

int main() {
    int n;
    while (cin >> n) {
        int len = pow(3, n);      // 총 길이
        string result(len, '-');  // 전부 '-'로 초기화

        makeCantor(result, 0, len);  // 가운데 공백을 넣기 위한 재귀

        cout << result << '\n';   // 최종 문자열 출력
    }
    return 0;
}

File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.java
import java.util.Scanner;

public class Main {

    public static void cantor(char[] arr, int start, int end) {
        if (end - start < 3) return;

        int third = (end - start) / 3;
        // 가운데 부분 공백 처리
        for (int i = start + third; i < start + 2 * third; i++) {
            arr[i] = ' ';
        }

        // 좌우 구간 재귀 처리
        cantor(arr, start, start + third);
        cantor(arr, start + 2 * third, end);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (sc.hasNextInt()) {
            int n = sc.nextInt();
            int size = (int) Math.pow(3, n);
            char[] arr = new char[size];

            // 초기화: 전부 '-'
            for (int i = 0; i < size; i++) {
                arr[i] = '-';
            }

            cantor(arr, 0, size);

            // 출력
            System.out.println(new String(arr));
        }
    }
}
